### PR TITLE
Migrate DirectCSI drives/volumes at install time

### PR DESCRIPTION
**Pull request by balamurugana**
Signed-off-by: Bala.FA <bala@minio.io>

### CODE REVIEW

**Review by cniackz — APPROVED** [review]
Hello Bala,

1. I like this PR because you moved all the logic from `cmd/kubectl-directpv/migrate.go` to `pkg/installer/migrate.go` which is good.
2. I don't see where the YAML output is being performed, rather I see the creation of volumes and drives on the fly, is that the new intention?
3. While testing this PR with my previous steps: https://github.com/cniackz/public/wiki/How-to-test-PR-652 I noticed that I cannot longer install legacy DirectPV because I see below error in direct-csi pod:
```
Readiness probe failed: Get "https://10.88.0.3:30443/healthz": dial tcp 10.88.0.3:30443: connect: connection refused
```
Anyway, maybe something I am missing or some unrelated issue, not sure. If by any chance you have documented the steps on how to test this PR, please share with me and thank you!

File: .github/workflows/functests.yml [hunk 1]
```diff
@@ -73,3 +73,12 @@ jobs:
       - name: Run tests
         run: |
           functests/run-tests.sh
+
+      - name: Run migration tests with DirectCSI v3.2.0
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply .github/workflows/functests.yml:77]: done

File: cmd/directpv/legacy-node-server.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		return startLegacyNodeServer(c.Context())
+	},
+}
+
+func startLegacyNodeServer(ctx context.Context) error {
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
+	defer cancel()
+
+	idServer, err := pkgidentity.NewServer("direct-csi-min-io", Version, map[string]string{})
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply cmd/directpv/legacy-node-server.go:44]: done

File: pkg/installer/daemonset.go [hunk 3]
```diff
@@ -218,61 +319,66 @@ func doCreateDaemonset(ctx context.Context, args *Args) error {
...
-					},
-				},
-				Spec: podSpec,
-			},
+	_, err = io.WriteString(args.auditWriter, mustGetYAML(daemonset))
+	return err
+}
+
+func doCreateLegacyDaemonset(ctx context.Context, args *Args) error {
+	if err := createOrUpdateLegacyNodeAPIServerSecrets(ctx, args); err != nil {
```
[the remark this answers is not in the document]
Comment (balamurugana) [reply pkg/installer/daemonset.go:350]: done

**Comment by balamurugana** [comment]
@cniackz 

> Hello Bala,
> 
>     1. I like this PR because you moved all the logic from `cmd/kubectl-directpv/migrate.go` to `pkg/installer/migrate.go` which is good.
> 
>     2. I don't see where the YAML output is being performed, rather I see the creation of volumes and drives on the fly, is that the new intention?
> 

As we create drives and volumes directly in migration, we don't need interim YAML.

>     3. While testing this PR with my previous steps: https://github.com/cniackz/public/wiki/How-to-test-PR-652 I noticed that I cannot longer install legacy DirectPV because I see below error in direct-csi pod:
> 
> 
> ```
> Readiness probe failed: Get "https://10.88.0.3:30443/healthz": dial tcp 10.88.0.3:30443: connect: connection refused
> ```
> 
> Anyway, maybe something I am missing or some unrelated issue, not sure. If by any chance you have documented the steps on how to test this PR, please share with me and thank you!

Testing this PR is simpler now. Below steps we need
1. install direct-csi v3.2.1 or v2.0.9
2. format drives and apply `minio.yaml` to create minio cluster which internally uses direct-csi to create volumes
3. once minio cluster is running, uninstall direct-csi
4. install directpv having migration by this PR
5. now we have to see migrated drives and volumes
6. delete minio cluster and create again by `minio.yaml`
7. now it should work without any issues.

With the above steps, this PR has migration functional tests as well and runs by github workflow CI.

File: pkg/installer/migrate.go [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+			continue
+		}
+
+		name := result.Volume.Name
+		nodeID := directpvtypes.NodeID(result.Volume.Status.NodeName)
+		driveID := directpvtypes.DriveID(fsuuid)
+		driveName := directpvtypes.DriveName(result.Volume.Labels["direct.csi.min.io/drive-path"])
+		size := result.Volume.Status.TotalCapacity
+
+		volume := types.NewVolume(
```
Comment (Praveenrajmani) [inline pkg/installer/migrate.go:224]: the migrated volume will always be in pending state here. We can make it to Ready state if the stagingtargetpath and containerpath isn't empty.

**Comment by Praveenrajmani** [comment]
```

File: pkg/installer/daemonset.go [hunk 5]
```diff
@@ -79,33 +82,32 @@ func createOrUpdateNodeAPIServerSecrets(ctx context.Context, args *Args) error {
...
+func getVolumesAndMounts(pluginSocketDir, certsSecretName string) (volumes []corev1.Volume, volumeMounts []corev1.VolumeMount) {
+	volumes = []corev1.Volume{
+		newHostPathVolume(volumeNameSocketDir, pluginSocketDir),
 		newHostPathVolume(volumeNameMountpointDir, kubeletDirPath+"/pods"),
 		newHostPathVolume(volumeNameRegistrationDir, kubeletDirPath+"/plugins_registry"),
 		newHostPathVolume(volumeNamePluginDir, kubeletDirPath+"/plugins"),
 		newHostPathVolume(volumeNameAppRootDir, appRootDir),
 		newHostPathVolume(volumeNameSysDir, volumePathSysDir),
 		newHostPathVolume(volumeNameDevDir, volumePathDevDir),
 		newHostPathVolume(volumeNameRunUdevData, volumePathRunUdevData),
```
Comment (Praveenrajmani) [inline pkg/installer/daemonset.go:107]: we also need to mount `/var/lib/direct-csi/` in legacy mode. Else, the drive controller cannot `remove` the drive (kubectl directpv remove). ie, it cannot umount the primary mount `/var/lib/direct-csi/` as the path is not mounted to the container.

The following error will be seen while umounting,

```log
E1215 07:07:59.096571  592613 listener.go:181] no such file or directory
```

It worked after applying the following diffs

```diff
diff --git a/pkg/consts/consts.go b/pkg/consts/consts.go
index 946263c1..01e19d93 100644
--- a/pkg/consts/consts.go
+++ b/pkg/consts/consts.go
@@ -64,6 +64,9 @@ const (
        // AppRootDir is application root directory.
        AppRootDir = "/var/lib/" + AppName
 
+       // LegacyAppRootDir denotes the legacy root directory.
+       LegacyAppRootDir = "/var/lib/direct-csi"
+
        // UdevDataDir is Udev data directory.
        UdevDataDir = "/run/udev/data"
 
diff --git a/pkg/consts/consts.go.in b/pkg/consts/consts.go.in
index b424646c..7320535a 100644
--- a/pkg/consts/consts.go.in
+++ b/pkg/consts/consts.go.in
@@ -62,6 +62,9 @@ const (
        // AppRootDir is application root directory.
        AppRootDir = "/var/lib/" + AppName
 
+       // LegacyAppRootDir denotes the legacy root directory.
+       LegacyAppRootDir = "/var/lib/direct-csi"
+
        // UdevDataDir is Udev data directory.
        UdevDataDir = "/run/udev/data"
```

```diff
diff --git a/pkg/installer/daemonset.go b/pkg/installer/daemonset.go
index a46dd398..ba9a947f 100644
--- a/pkg/installer/daemonset.go
+++ b/pkg/installer/daemonset.go
@@ -40,7 +40,9 @@ const (
        volumeNameRegistrationDir          = "registration-dir"
        volumeNamePluginDir                = "plugins-dir"
        volumeNameAppRootDir               = consts.AppName + "-common-root"
+       volumeNameLegacyAppRootDir         = "direct-csi-common-root"
        appRootDir                         = consts.AppRootDir + "/"
+       legacyAppRootDir                   = consts.LegacyAppRootDir + "/"
        volumeNameSysDir                   = "sysfs"
        volumeNameDevDir                   = "devfs"
        volumePathDevDir                   = "/dev"
@@ -102,6 +104,7 @@ func getVolumesAndMounts(pluginSocketDir, certsSecretName string) (volumes []cor
                newHostPathVolume(volumeNameRegistrationDir, kubeletDirPath+"/plugins_registry"),
                newHostPathVolume(volumeNamePluginDir, kubeletDirPath+"/plugins"),
                newHostPathVolume(volumeNameAppRootDir, appRootDir),
+               newHostPathVolume(volumeNameLegacyAppRootDir, legacyAppRootDir),
                newHostPathVolume(volumeNameSysDir, volumePathSysDir),
                newHostPathVolume(volumeNameDevDir, volumePathDevDir),
                newHostPathVolume(volumeNameRunUdevData, volumePathRunUdevData),
@@ -112,6 +115,7 @@ func getVolumesAndMounts(pluginSocketDir, certsSecretName string) (volumes []cor
                newVolumeMount(volumeNameMountpointDir, kubeletDirPath+"/pods", corev1.MountPropagationBidirectional, false),
                newVolumeMount(volumeNamePluginDir, kubeletDirPath+"/plugins", corev1.MountPropagationBidirectional, false),
                newVolumeMount(volumeNameAppRootDir, appRootDir, corev1.MountPropagationBidirectional, false),
+               newVolumeMount(volumeNameLegacyAppRootDir, legacyAppRootDir, corev1.MountPropagationBidirectional, false),
                newVolumeMount(volumeNameSysDir, volumePathSysDir, corev1.MountPropagationBidirectional, true),
                newVolumeMount(volumeNameDevDir, volumePathDevDir, corev1.MountPropagationHostToContainer, true),
                newVolumeMount(volumeNameRunUdevData, volumePathRunUdevData, corev1.MountPropagationBidirectional, true),

```

**Comment by Praveenrajmani** [comment]
Tested and LGTM otherwise

**Review by Praveenrajmani — APPROVED** [review]

**Pull request merged**